src/angular/components/sso.component.ts

```diff
@@ -0,0 +1,133 @@
+
+import { Utils } from '../../misc/utils';
+
+import { AuthResult } from '../../models/domain/authResult';
+
+export class SsoComponent {
+    identifier: string;
+    loggingIn = false;
+
+    formPromise: Promise<AuthResult>;
+    onSuccessfulLogin: () => Promise<any>;
+    onSuccessfulLoginNavigate: () => Promise<any>;
+    onSuccessfulLoginTwoFactorNavigate: () => Promise<any>;
+    onSuccessfulLoginChangePasswordNavigate: () => Promise<any>;
+
+    protected twoFactorRoute = '2fa';
+    protected successRoute = 'lock';
+    protected changePasswordRoute = 'change-password';
+    protected redirectUri: string;
+
+    constructor(protected authService: AuthService, protected router: Router,
+        protected i18nService: I18nService, protected route: ActivatedRoute,
+        protected storageService: StorageService, protected stateService: StateService,
+        protected platformUtilsService: PlatformUtilsService, protected apiService: ApiService,
+        protected cryptoFunctionService: CryptoFunctionService,
+        protected passwordGenerationService: PasswordGenerationService) { }
+
+    async ngOnInit() {
+        const queryParamsSub = this.route.queryParams.subscribe(async (qParams) => {
+            if (qParams.code != null && qParams.state != null) {
+                const codeVerifier = await this.storageService.get<string>(ConstantsService.ssoCodeVerifierKey);
+                const state = await this.storageService.get<string>(ConstantsService.ssoStateKey);
+                await this.storageService.remove(ConstantsService.ssoCodeVerifierKey);
+                await this.storageService.remove(ConstantsService.ssoStateKey);
+                if (qParams.code != null && codeVerifier != null && state != null && state === qParams.state) {
+                    await this.logIn(qParams.code, codeVerifier);
+                }
+            }
+            if (queryParamsSub != null) {
+                queryParamsSub.unsubscribe();
+            }
+        });
+    }
+
+    async submit() {
+        const passwordOptions: any = {
+            type: 'password',
+            length: 64,
+            uppercase: true,
+            lowercase: true,
+            numbers: true,
+            special: false,
+        };
+        const state = await this.passwordGenerationService.generatePassword(passwordOptions);
+        const codeVerifier = await this.passwordGenerationService.generatePassword(passwordOptions);
+        const codeVerifierHash = await this.cryptoFunctionService.hash(codeVerifier, 'sha256');
+        const codeChallenge = Utils.fromBufferToUrlB64(codeVerifierHash);
+
+        await this.storageService.save(ConstantsService.ssoCodeVerifierKey, codeVerifier);
+        await this.storageService.save(ConstantsService.ssoStateKey, state);
+
+        const authorizeUrl = this.apiService.identityBaseUrl + '/connect/authorize?' +
+            'client_id=web&redirect_uri=' + this.redirectUri + '&' +
```

Comment:
Does `this.redirectUri` need to be URL encoded?

Reply:
Is URL encoding done as a safety precaution? Dumb question - if a string component is encoded, does it need to be decoded once received or does it all happen auto-magically?

Reply:
Encoding must be done or you lose the value of certain characters. Decoding happens automatically.

Reply:
Updated